File: android/src/tibi/fruity/AndroidLauncher.java
package tibi.fruity;

import java.io.IOException;

import android.content.res.AssetFileDescriptor;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.media.PlaybackParams;
import android.os.Bundle;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.backends.android.AndroidApplication;
import com.badlogic.gdx.backends.android.AndroidApplicationConfiguration;
import com.badlogic.gdx.backends.android.AndroidFileHandle;

public class AndroidLauncher extends AndroidApplication {

	private final MusicPlayer player = new MusicPlayerAndroid();

	@Override
	protected void onCreate (Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		AndroidApplicationConfiguration config = new AndroidApplicationConfiguration();
		config.useAccelerometer = false;
		config.useCompass = false;

		initialize(new FruityFrankGame(player), config);
	}

	@Override protected void onPause() {
		super.onPause();
		player.release();
	}


	public class MusicPlayerAndroid implements MusicPlayer {

		private MediaPlayer player;

		@Override public void play(float speedFactor) {
			try {
				release();
				player = new MediaPlayer();
				AndroidFileHandle file = (AndroidFileHandle) Gdx.files.internal("level 1.mid");
				AssetFileDescriptor descriptor = file.getAssetFileDescriptor();
				player.setDataSource(descriptor.getFileDescriptor(),
				                     descriptor.getStartOffset(), descriptor.getLength());
				player.setAudioStreamType(AudioManager.STREAM_MUSIC);
				PlaybackParams params = new PlaybackParams();
				params.setSpeed(speedFactor);
				player.setPlaybackParams(params);
				player.setLooping(true);
				player.prepare();
				player.start();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}

		@Override public void release() {
			if (player != null) {
				player.release();
			}
		}
	}
}

File: android/src/tibi/fruity/AndroidLauncher.kt
package tibi.fruity

import android.media.AudioAttributes
import android.media.MediaPlayer
import android.media.PlaybackParams
import android.os.Bundle
import com.badlogic.gdx.backends.android.AndroidApplication
import com.badlogic.gdx.backends.android.AndroidApplicationConfiguration
import com.badlogic.gdx.backends.android.AndroidFileHandle
import java.io.IOException

class AndroidLauncher : AndroidApplication() {

    private val player = MusicPlayerAndroid()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val config = AndroidApplicationConfiguration()
        config.useAccelerometer = false
        config.useCompass = false
        config.useImmersiveMode = true
        initialize(FruityFrankGame(player), config)
    }

    override fun onPause() {
        super.onPause()
        player.pause(true)
    }

    override fun onResume() {
        super.onResume()
        player.pause(false)
    }

    inner class MusicPlayerAndroid : MusicPlayer {

        private var mplayer: MediaPlayer? = null
        var fileNamePrefix: String? = null
        var speedFactor: Float? = null

        override fun play(fileNamePrefix: String, speedFactor: Float) {
            this.fileNamePrefix = fileNamePrefix
            this.speedFactor = speedFactor
            release()
            start()
        }

        private fun start() {
            try {
                if (fileNamePrefix == null) {
                    return
                }
                val player = MediaPlayer()
                val file = findFile(fileNamePrefix!!) as AndroidFileHandle
                val descriptor = file.assetFileDescriptor
                player.setDataSource(descriptor.fileDescriptor,
                        descriptor.startOffset, descriptor.length)
                player.setAudioAttributes(AudioAttributes
                               .Builder()
                               .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                               .build())
                val params = PlaybackParams()
                params.speed = speedFactor!!
                player.playbackParams = params
                player.isLooping = true
                player.prepare()
                player.start()
                mplayer = player
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }

        override fun pause(value: Boolean) {
            if (value) mplayer?.pause()
            else {
                if (mplayer == null) start()
                else mplayer?.start()
            }
        }

        override fun release() {
            mplayer?.release()
            mplayer = null
        }
    }
}
